build_tools/android/run_benchmarks.py

```diff
@@ -0,0 +1,460 @@
+
+
+def run_python_model_benchmark_suite(device_info,
+                                     root_build_dir,
+                                     model_benchmark_dirs,
+                                     benchmark_tool,
+                                     verbose=False):
+  """Runs all model benchmarks on the Android device and report results.
+
+  Args:
+  - device_info: an AndroidDeviceInfo object.
+  - root_build_dir: the root build directory.
+  - model_benchmark_dirs: a list of model benchmark directories.
+  - benchmark_tool: the path to the benchmark tool.
+
+  Returns:
+  - A dictionary containing maps from benchmark names to the time (ms).
+  """
+  # Push the benchmark tool to the Android device first.
+  android_tool_path = adb_push_to_tmp_dir(benchmark_tool,
+                                          relative_dir="tools",
+                                          verbose=verbose)
+
+  model_root_dir = os.path.join(root_build_dir, BENCHMARK_SUITE_REL_PATH,
+                                PYTON_MODEL_REL_PATH)
+
+  results = {}
+
+  # Push all model artifacts to the device and run them.
+  for model_benchmark_dir in model_benchmark_dirs:
+    benchmark_name = compose_benchmark_name(device_info, root_build_dir,
+                                            model_benchmark_dir)
+    print(f"--> benchmark: {benchmark_name} <--")
+    android_relative_dir = os.path.relpath(model_benchmark_dir, model_root_dir)
+    adb_push_to_tmp_dir(os.path.join(model_benchmark_dir, MODEL_VMFB_NAME),
+                        android_relative_dir,
+                        verbose=verbose)
+    android_flagfile_path = adb_push_to_tmp_dir(os.path.join(
+        model_benchmark_dir, MODEL_FLAGFILE_NAME),
+                                                android_relative_dir,
+                                                verbose=verbose)
+
+    cmd = [
+        android_tool_path,
+        f"--flagfile={android_flagfile_path}",
+        f"--benchmark_repetitions={BENCHMARK_REPETITIONS}",
+        "--benchmark_format=json",
+    ]
+    resultjson = adb_execute_in_path(cmd, android_relative_dir, verbose=verbose)
+
+    print(resultjson)
+    resultjson = json.loads(resultjson)
+
+    if benchmark_name in results:
+      raise ValueError(f"Duplicated benchmark name: {benchmark_name}")
+
+    avg_time = None
+    for benchmark in resultjson["benchmarks"]:
+      if benchmark["name"].endswith("real_time_mean"):
+        if benchmark["time_unit"] != "ms":
+          raise ValueError(
+              f"Expected ms as time unit but found {benchmark['time_unit']}")
+        avg_time = int(benchmark["real_time"])
+        break
+    if avg_time is None:
+      raise ValueError(f"Cannot find average time")
+
+    results[benchmark_name] = avg_time
+
+  return results
+
+
+def parse_arguments():
+  """Parses command-line options."""
+
+  def check_dir_path(path):
+    if os.path.isdir(path):
+      return path
+    else:
+      raise NotADirectoryError(path)
```

Comment:
Only ArgumentTypeError, TypeError, or ValueError should be thrown in callables passed to `type` (https://docs.python.org/3/library/argparse.html#type). Seems like ArgumentTypeError would be appropriate for both of these?

Reply:
Oh didn't know about that. Thanks!